google/cloud/storage/retry.py

```diff
@@ -125,6 +125,8 @@ def is_etag_in_json(data):
 
     Indended for use on calls with relatively short JSON payloads."""
     try:
+        if isinstance(data, dict) and data.get("etag"):
```

Comment:
It's a bit disturbing that this can be called with both dict and string data. Is there a way to fix the call sites so that we are passing in a consistent kind of argument to this function, rather than accepting both?

Reply:
If we *do* have this method take both a JSON-encoded string and a `dict`, then we should factor it such that the JSON-encoded string is decoded to a `dict` *first*, leaving the test the same. Also, we should update the docstring to indicate that both types are allowed:

```python
def is_etag_in_json(data):
    """Return True if an etag is contained in the JSON body.

    :type data: str or dict
    :param data: A string containing a JSON-encoded dict, or the dict itsefl.

    Intended for use on calls with relatively short JSON payloads."""
    if isinstance(data, six.text_type):
        try:
            data = json.loads(data)
        # Though this method should only be called when a JSON body is expected,
        # the retry policy should be robust to unexpected payloads.
        # In Python 3 a JSONDecodeError is possible, but it is a subclass of ValueError.
        except (ValueError, TypeError):
            return False
    return "etag" in data
```

Reply:
The change in 73a9802 LGTM, assuming @tritone's concern above is addressed.

Reply:
Thanks @tseaver @tritone for the reviews. I understand it can be quite confusing. Given that retries are configurable, there are cases where `data` is type string or type dict. In our discussion, @andrewsg pointed out many cases `data` is a string containing a JSON-encoded dict. Meanwhile, the callers of this function in our library pass `data` as a dict, for example:
- [`client._put_resource`](https://github.com/googleapis/python-storage/blob/master/google/cloud/storage/client.py#L490) where `data` is type dict (called by `bucket.set_iam_policy` , `hmac_key.update`, `bucket.update`, `blob.update`)
- [`client._patch_resource`](https://github.com/googleapis/python-storage/blob/1ae766ed3e1313439c5a8698562c02d1a0344ff3/google/cloud/storage/client.py#L419) where `data` is type dict (called by `bucket.patch`, `blob.patch`)

Another option would be modifying all the `data` types in the above-mentioned library methods?

Reply:
@cojenco 

>  In our discussion, @andrewsg pointed out many cases data is a string containing a JSON-encoded dict. 

I'm not sure where `data` is going to be a string in cases where `if_etag_in_json` is called.

A quick survey of usage:

- Outside of tests, `is_etag_in_json` is used only as the predicate for the `DEFAULT_RETRY_IF_ETAG_IN_JSON` retry policy:

```bash
$ git grep -B 1 is_etag_in_json google/
google/cloud/storage/retry.py-
google/cloud/storage/retry.py:def is_etag_in_json(data):
--
google/cloud/storage/retry.py-DEFAULT_RETRY_IF_ETAG_IN_JSON = ConditionalRetryPolicy(
google/cloud/storage/retry.py:    DEFAULT_RETRY, is_etag_in_json, ["data"]
```
The `DEFAULT_RETRY_IF_ETAG_IN_JSON` policy is used in exactly three places:

```bash
$ git grep DEFAULT_RETRY_IF_ETAG_IN_JSON google/ | grep -v import
google/cloud/storage/blob.py:        retry=DEFAULT_RETRY_IF_ETAG_IN_JSON,
google/cloud/storage/bucket.py:        retry=DEFAULT_RETRY_IF_ETAG_IN_JSON,
google/cloud/storage/hmac_key.py:    def update(self, timeout=_DEFAULT_TIMEOUT, retry=DEFAULT_RETRY_IF_ETAG_IN_JSON):
google/cloud/storage/retry.py:DEFAULT_RETRY_IF_ETAG_IN_JSON = ConditionalRetryPolicy(
```

which turn out to be:

- `Blob.set_iam_policy` (passes a `dict` to `client._put_resource`)
- `Bucket.set_iam_policy` (passes a `dict` to `client._put_resource`)
- `HMACKeyMetadata.update` (passes a `dict` to `client._put_resource`)

Although `Bucket.patch`, `Blob.patch`, `Bucket.update`, and `Blob.update` do not use the `DEFAULT_RETRY_IF_ETAG_IN_JSON` policy, they all pass `dict` to `client._patch_resource`.

We do expose `DEFAULT_RETRY_IF_ETAG_IN_JSON` as a public symbol in the public `retry` module, so I guess there could be cases out in the wild which use it for some other API calls, but I'm doubtful any such hypothetical users are actually getting real retries done.

Reply:
FWIW, I just tweaked the implementation of `is_etag_in_json` as follows:

```python
def is_etag_in_json(data):
    """Return True if an etag is contained in the JSON body.

    Indended for use on calls with relatively short JSON payloads."""
    return data.get("etag") is not None
```

and all systests pass.

Reply:
Digging further:  since PRs #436, #441, and #443,  `Client._patch_resource`, `Client._put_resource`, and `Client._post_resource` are the only three places where the `Connection.api_request` method is called with a `data` argument.  Therefore their callers are the only ones which could potentially use the `DEFAULT_RETRY_IF_ETAG_IN_JSON` policy.

There are a few cases where `Client._post_resource` is called with `data` passed as `None`:  otherwise, `data` is always a `dict` instance.  So, I think we could simplify `is_etag_in_json` to just:

```python
    return data is not None and "etag" in data
```

The various upload mechanisms might still end up passing in `data` as bytes at a lower level, but retrying based on an `ETag` is never sensible for them:  they all create new versions of the blob, and `ETag` is not a writable property.

Reply:
Thanks for digging in and clarifying the issue. That makes lots of sense. Knowing that the library callers are passing in data as `dict` or `None`, I'm not worried about revising the predicate. My (no longer) biggest concern was `DEFAULT_RETRY_IF_ETAG_IN_JSON` exposed and configurable in the public retry module. I like the idea of simplifying the predicate. Updating the changes, thanks!